### PR TITLE
Replacement of internal RegEx with PCRE2

### PR DESCRIPTION
**Pull request by leezh**
Turns out PCRE2 was not as heavy as I thought and was fairly easy to integrate into Godot. This is the minimal feature-parity-ish commit. Plans to introduce serialised compilations and JIT speed-ups would come separately later.

~~**That said, discussion is needed for `modules/regex/SCsub`.** Due to how `wchar_t` can be 32 bit on *nix but 16-bit on Windows, it is necessary to know at compile-time its width. Unfortunately, I'm not sure the tests in `modules/regex/SCsub` is the right place (or even the right method).~~

~~If someone more familiar with SCons can help me out, that'd be great.~~

## Notable minor breaking changes:

* The pattern and replacement matching behaviour has been changed purely due to the nature of switching to a standards-compliant library.
* One mistake in the previous behaviour was that named groups didn't have a number. This has been corrected.
* As names are actually just an alias of numbered groups, `RegExMatch::get_name_dict()` has been removed. Also, for clarity, `RegExMatch::get_group_array()` has been renamed.
* Due the lack of a suitable equivalent in PCRE2, `RegExMatch::expand()` was removed.

### CODE REVIEW

**Comment by tagcup** [comment]
re2 or hyperscan perform much better than pcre, so are there any good reasons that I'm missing for choosing pcre over them? Its performance with JIT becomes comparable to them, apparently, but JIT may not be available in all platforms. Also re2 has a linear-time processing (which comes at the cost of dropping back-references).

**Comment by leezh** [comment]
The biggest motivation for picking PCRE is because I can easily choose between 16-bit character or 32-bit character strings at compile time. In doing so I can get it to play nice with Godot's built-in String type rather than converting back and forth between UTF-8.

**Comment by tagcup** [comment]
Ah, does this String class use UCS-4?

**Comment by leezh** [comment]
Annoyingly, not exactly. It uses `wchar_t` which is UCS-2 on Windows but
UCS-4 on *Nix systems.

Also, forced pushed some clang formatting fixes I forgot to do.

**Comment by tagcup** [comment]
I'm guessing godot's string class isn't using utf8 for historical reasons; maybe someone knows if there are any plans for a switch?

**Comment by leezh** [comment]
No, there's no plans to switch away from `wchar_t`.

As discussed with reduz on IRC, I've moved away from SCons trying to detect
the width of `wchar_t` as doing a test compile can be unreliable at times.
Instead it'll be done during run-time, with hope that optimisers would get
rid of the unneeded parts of the code.

File: modules/regex/regex.cpp [hunk 1]
```diff
@@ -29,1479 +29,493 @@
...
+	}
 }
 
 void RegEx::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("clear"), &RegEx::clear);
 	ClassDB::bind_method(D_METHOD("compile", "pattern"), &RegEx::compile);
-	ClassDB::bind_method(D_METHOD("search:RegExMatch", "text", "start", "end"), &RegEx::search, DEFVAL(0), DEFVAL(-1));
-	ClassDB::bind_method(D_METHOD("sub", "text", "replacement", "all", "start", "end"), &RegEx::sub, DEFVAL(false), DEFVAL(0), DEFVAL(-1));
+	ClassDB::bind_method(D_METHOD("search:RegExMatch", "subject", "offset", "end"), &RegEx::search, DEFVAL(0), DEFVAL(-1));
```
Comment (neikeq) [inline modules/regex/regex.cpp]: Starting today, the `:type` suffix must not be used anymore, except for virtual method binds.

Comment (leezh) [inline modules/regex/regex.cpp]: Done! Also fixed the merge conflicts.

**Comment by akien-mga** [comment]
There seems to be a linking issue on Windows (at least old mingw version on Ubuntu 14.04): https://travis-ci.org/godotengine/godot/jobs/263722203

**Comment by leezh** [comment]
Yeah, realised it was missing the `-DPCRE_STATIC` flag and a few others. It should be fixed now.

File: modules/regex/SCsub [hunk 2]
```diff
@@ -1,7 +1,53 @@
...
+
+def pcre2_builtin(width):
+    env_pcre2 = env_modules.Clone()
+    env_pcre2["OBJSUFFIX"] = "_" + width + env_pcre2["OBJSUFFIX"]
+    env_pcre2.Append(CPPPATH=[thirdparty_dir])
+    env_pcre2.add_source_files(env.modules_sources, thirdparty_sources)
+    env_pcre2.Append(CPPFLAGS=thirdparty_flags)
+    env_pcre2.Append(CPPFLAGS=["-DPCRE2_CODE_UNIT_WIDTH=" + width])
+
+if (env['builtin_pcre2'] != 'no'):
```
Comment (akien-mga) [inline modules/regex/SCsub]: Most of the above should go inside the `if` actually, when building against the system pcre2 (typically on Linux), we only want to build the module files themselves, not reference anything in `thirdparty_dir`.

Could you also add the `platform/x11/detect.py` detection of pcre2 with pkgconfig, as done for other opt-out thirdparty deps?

Comment (akien-mga) [inline modules/regex/SCsub]: (If you're not on Linux, I could likely add it myself after merge if need be)

Comment (leezh) [inline modules/regex/SCsub]: So I've moved as much as possible into the `if` block to allow external linking.

On X11, `builtin_pcre2` is still by default set to yes, as it's not available on Ubuntu Trusty (and thus Travis-CI). Is there a way to automatically select yes/no if the native lib is not-found/found?

**Pull request merged**